### DISCUSSION TITLE
Ubuntu systems showing lots of core packages as non-compliant.

### DISCUSSION THREAD

**Discussion by ppanon2022** [uyuni-project/uyuni#8502] [General]
I've noticed that a lot of our Ubuntu client systems are now showing a lot of core packages as non-compliant. 
On one client system, running apt list --upgradeable (and afterwards apt update && apt upgrade)  showed many outstanding updates from the "Non-compliant" list. Forcing a package list update from the Uyuni console doesn't change the result. Applying the updates with apt upgrade on the client made those packages disappear from the non-compliant list. Is anybody else experiencing the same problem?

In our case, I'm not exactly sure what caused it. We skipped upgrading to 2024.01 and went straight from 2023.12 to 2024.02. We also had an issue with some package syncs failing due to needing to increase available storage. So I can't know which of those changes/issues caused the change in behaviour. However I'm not seeing the same problems with rpm based distros like CentOS, OpenSUSE, or Redhat on two different servers. So I'm leaning to thinking it's likely due to a change in the package comparison/matching algorithm for the on-compliant list. Was there any change affecting that in 2024.01 or 2024.02?

This is very concerning because it means needed Ubuntu updates are not being deployed.

**Comment by ppanon2022** [comment]
I ran a reindex database with Uyuni shut down, Then when I bought it back up, I ran a package list refresh, however that seemed to make no difference in the long list of packages showing up as non-compliant.

**Comment by akupreev** [comment]
@ppanon2022 Hello. Tell me, did you manage to fix this problem?

**Reply by ppanon2022** [reply]
Well, we updated to 2024.05 and also have done a couple of patching cycles since, and that seems to have helped. Also, during the upgrade to 2024.05, we noticed (for a different reason) that some spacewalk-backend* packages weren't getting updated and we had to do some work to resolve that (some package dependencies and old deprecated packages that needed cleaning up), so perhaps a version mismatch between the front and back ends was a factor. Or perhaps the schema had changed but some of the queries were from the older backend packages and no longer worked properly after the schema change.

**Reply by akupreev** [reply]
Good afternoon, my friend.
Do I understand correctly that:
1) You have installed patches on Uyuni Server (zypper patch)
2) Updated all spacewalk-backend* packages - Tell me, what version of the packages are they now?

My Version of Uyuni Server:
Repository : uyuni-server-stable
Name : Uyuni-Server-release
Version : 2024.05-230900.217.1.uyuni3
Arch : x86_64
Vendor : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level : Level 3
Installed Size : 1.4 KiB
Installed : Yes
Status : up-to-date
Source package : Uyuni-Server-release-2024.05-230900.217.1.uyuni3.src
Summary : Uyuni Server

**Reply by ppanon2022** [reply]
The thing is that Uyuni-Server-release isn't what you need to look at because it said 2024.05 was installed before as well. What you need to check is specifically the version of the spacewalk-backend* packages. They should all be the 5.0.6 version.

\# rpm -q -a | fgrep spacewalk-backend | sort
spacewalk-backend-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-app-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-applet-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-iss-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-iss-export-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-package-push-server-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-server-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-sql-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-sql-postgresql-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-tools-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-xml-export-libs-5.0.6-230900.1.4.uyuni3.noarch
spacewalk-backend-xmlrpc-5.0.6-230900.1.4.uyuni3.noarch

If they aren't that release then you need to do a zypper refresh && zypper update and see what packages are being held back. Then you need to zypper install any held back spacewalk-backend packages and resolve any dependency conflicts. See https://github.com/uyuni-project/uyuni/issues/8937

**Comment by ppanon2022** [comment]
I think one other thing is that the non-compliant list used to only show packages which did not have newer packages with the same name in any of the assigned channels (i.e. orphaned packages which were either manually copied and installed, or which were installed by adding a PPA on the system without any equivalent in an Uyuni channel/repo to replace it). Perhaps it also included packages which were in an assigned channel but had a newer version than the latest available in the channel - for example, when using LCM projects with a cutoff date filter to ensure package consistency between groups of  systems that are used to test updates in dev/sandbox, then QA/staging, prior to production deployment, new systems that are patched as part of an installation that occurred after the last version cutoff date for the channel could have a newer version of some packages issued after the cutoff date. That situation could also be aggravated if you're only applying security updates, but the Ubuntu installer applied all updates, including many bug-fix updates that aren't in the channel. I expect that means what probably helped the most in our case was running a patching cycle after that update to bring systems into consistency with the LCM channels.

Now the non-compliant list also seems to include all the packages which need to be updated. I understand why you would want to highlight down-level (update pending) versions as non-compliant, but it makes it much harder to identify which packages are from a 3rd party PPA with no channel equivalent because that info can be a needle buried in the haystack of packages needing updating. Perhaps there should be an additional tab (perhaps Unsupported?) with contents closer to the older Non-Compliant tab, listing packages for which no subscribed channel has a package of the same name but different version?

**Reply by rjmateus** [reply]
The non-compliant list should only show packages (name + version) that are installed and not part of any of the assigned software channel. 
Packages installed and that exist in the channels and have a new version are not considered non-compliant.

**Reply by ppanon2022** [reply]
That is not what we're seeing in our environments. For instance, here's the non-compliant list of an Ubuntu 22.04 system which was just imaged yesterday. This system was a fresh Ubuntu 22.04 install that applied all available updates and was subsequently subscribed to the LCM channels by the Activation Key, which includes only the security update channels. Then to see how it would affect that Non-Compliant list, I changed the system back to the non-LCM channels that include all updates and performed a Package List Refresh.

```
Package Name		Architecture		Installed
--		--		--
apt-2.4.11-X		amd64-deb		2/16/2024
apt-utils-2.4.11-X		amd64-deb		2/16/2024
bind9-dnsutils-9.18.18-0ubuntu0.22.04.2		amd64-deb		2/16/2024
bind9-host-9.18.18-0ubuntu0.22.04.2		amd64-deb		2/16/2024
bind9-libs-9.18.18-0ubuntu0.22.04.2		amd64-deb		2/16/2024
cloud-init-23.3.3-0ubuntu0~22.04.1		all-deb		2/16/2024
coreutils-8.32-4.1ubuntu1.1		amd64-deb		2/16/2024
dpkg-1.21.1ubuntu2.2-X		amd64-deb		2/16/2024
firmware-sof-signed-2.0-1ubuntu4.7		all-deb		Yesterday at 16:26
libapt-pkg6.0-2.4.11-X		amd64-deb		2/16/2024
libgpgme11-1.16.0-1.2ubuntu4.1		amd64-deb		2/16/2024
libldap-common-2.5.16+dfsg-0ubuntu0.22.04.2		all-deb		2/16/2024
nessusagent-10.5.1-X		amd64-deb		Yesterday at 17:12
python3-update-manager-22.04.18-X		all-deb		2/16/2024
snapd-2.58+22.04.1-X		amd64-deb		2/16/2024
tcpdump-4.99.1-3ubuntu0.1		amd64-deb		2/16/2024
tzdata-2024a-0ubuntu0.22.04		all-deb		Yesterday at 16:30
ubuntu-advantage-tools-30~22.04-X		amd64-deb		2/16/2024
ubuntu-minimal-1.481.1-X		amd64-deb		2/16/2024
ubuntu-pro-client-l10n-30~22.04-X		amd64-deb		2/16/2024
ubuntu-server-1.481.1-X		amd64-deb		2/16/2024
ubuntu-server-minimal-1.481.1-X		amd64-deb		2/16/2024
ubuntu-standard-1.481.1-X		amd64-deb		2/16/2024
update-manager-core-22.04.18-X		all-deb		2/16/2024
update-notifier-common-3.192.54.6-X		all-deb		2/16/2024
vim-8.2.3995-1ubuntu2.16		amd64-deb		Yesterday at 16:30
vim-common-8.2.3995-1ubuntu2.16		all-deb		Yesterday at 16:30
vim-runtime-8.2.3995-1ubuntu2.16		all-deb		Yesterday at 16:30
vim-tiny-8.2.3995-1ubuntu2.16		amd64-deb		Yesterday at 16:30
xxd-8.2.3995-1ubuntu2.16		amd64-deb		Yesterday at 16:30
```

The upgradeable packages list is empty. 

The software updates Main channel includes all the latest updates and have the following bind packages, including newer ones than the three bind9 packages in the Non-Compliant list above.
```
bind9-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	Internet Domain Name Server
bind9-dev-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	Static Libraries and Headers used by BIND 9
bind9-dnsutils-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	Clients provided with BIND 9
bind9-doc-9.18.24-0ubuntu0.22.04.1:1.all-deb	Documentation for BIND 9
bind9-host-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	DNS Lookup Utility
bind9-libs-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	Shared Libraries used by BIND 9
bind9-utils-9.18.24-0ubuntu0.22.04.1:1.amd64-deb	Utilities for BIND 9
```

If I switch the system back to the LCM channels with only security updates prior to an earlier cutoff date filter and run another package list refresh, then the non-compliant list is much longer:

```
Package Name	Architecture	Installed	
apparmor-3.0.4-2ubuntu2.3	amd64-deb	2/16/2024	
apport-2.20.11-0ubuntu82.5	all-deb	2/16/2024	
apt-2.4.11-X	amd64-deb	2/16/2024	
apt-utils-2.4.11-X	amd64-deb	2/16/2024	
base-files-12ubuntu4.6-X	amd64-deb	2/16/2024	
cloud-init-23.3.3-0ubuntu0~22.04.1	all-deb	2/16/2024	
coreutils-8.32-4.1ubuntu1.1	amd64-deb	2/16/2024	
cryptsetup-2.4.3-1ubuntu1.2	amd64-deb	2/16/2024	
cryptsetup-bin-2.4.3-1ubuntu1.2	amd64-deb	2/16/2024	
cryptsetup-initramfs-2.4.3-1ubuntu1.2	all-deb	2/16/2024	
distro-info-1.1ubuntu0.2-X	amd64-deb	2/16/2024	
dmidecode-3.3-3ubuntu0.1	amd64-deb	2/16/2024	
dpkg-1.21.1ubuntu2.2-X	amd64-deb	2/16/2024	
firmware-sof-signed-2.0-1ubuntu4.7	all-deb	Yesterday at 16:26	
fwupd-1.7.9-1~22.04.3	amd64-deb	2/16/2024	
fwupd-signed-1.51.1~22.04.1+1.4-0ubuntu0.1	amd64-deb	2/16/2024	
grub2-common-2.06-2ubuntu7.2	amd64-deb	Yesterday at 16:25	
grub-common-2.06-2ubuntu7.2	amd64-deb	Yesterday at 16:25	
gzip-1.10-4ubuntu4.1	amd64-deb	2/16/2024	
initramfs-tools-0.140ubuntu13.4-X	all-deb	2/16/2024	
initramfs-tools-bin-0.140ubuntu13.4-X	amd64-deb	2/16/2024	
initramfs-tools-core-0.140ubuntu13.4-X	all-deb	2/16/2024	
iptables-1.8.7-1ubuntu5.2	amd64-deb	2/16/2024	
irqbalance-1.8.0-1ubuntu0.2	amd64-deb	2/16/2024	
isc-dhcp-client-4.4.1-2.3ubuntu2.4	amd64-deb	2/16/2024	
isc-dhcp-common-4.4.1-2.3ubuntu2.4	amd64-deb	2/16/2024	
kbd-2.3.0-3ubuntu4.22.04	amd64-deb	2/16/2024	
kpartx-0.8.8-1ubuntu1.22.04.4	amd64-deb	2/16/2024	
ldap-utils-2.5.18+dfsg-0ubuntu0.22.04.1	amd64-deb	Yesterday at 18:04	
libapparmor1-3.0.4-2ubuntu2.3	amd64-deb	2/16/2024	
libapt-pkg6.0-2.4.11-X	amd64-deb	2/16/2024	
libcryptsetup12-2.4.3-1ubuntu1.2	amd64-deb	2/16/2024	
libcups2-2.4.1op1-1ubuntu4.10	amd64-deb	Yesterday at 18:04	
libdrm2-2.4.113-2~ubuntu0.22.04.1	amd64-deb	2/16/2024	
libdrm-common-2.4.113-2~ubuntu0.22.04.1	all-deb	2/16/2024	
libfwupd2-1.7.9-1~22.04.3	amd64-deb	2/16/2024	
libfwupdplugin5-1.7.9-1~22.04.3	amd64-deb	2/16/2024	
libgpgme11-1.16.0-1.2ubuntu4.1	amd64-deb	2/16/2024	
libgstreamer1.0-0-1.20.3-0ubuntu1	amd64-deb	2/16/2024	
libip4tc2-1.8.7-1ubuntu5.2	amd64-deb	2/16/2024	
libip6tc2-1.8.7-1ubuntu5.2	amd64-deb	2/16/2024	
libipa-hbac0-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libldap-2.5-0-2.5.18+dfsg-0ubuntu0.22.04.1	amd64-deb	Yesterday at 18:04	
libmbim-glib4-1.28.0-1~ubuntu20.04.1	amd64-deb	2/16/2024	
libmbim-proxy-1.28.0-1~ubuntu20.04.1	amd64-deb	2/16/2024	
libmm-glib0-1.20.0-1~ubuntu22.04.3	amd64-deb	2/16/2024	
libnetplan0-0.106.1-7ubuntu0.22.04.4	amd64-deb	Yesterday at 16:30	
libnfsidmap1-2.6.1-1ubuntu1.2	amd64-deb	Yesterday at 18:04	
libnftables1-1.0.2-1ubuntu3	amd64-deb	2/16/2024	
libnss-sss-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libnss-systemd-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
libpam-sss-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libpam-systemd-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
libpython3.10-3.10.12-1~22.04.4	amd64-deb	Yesterday at 16:30	
libpython3.10-minimal-3.10.12-1~22.04.4	amd64-deb	Yesterday at 16:30	
libpython3.10-stdlib-3.10.12-1~22.04.4	amd64-deb	Yesterday at 16:30	
libqmi-glib5-1.32.0-1ubuntu0.22.04.1	amd64-deb	2/16/2024	
libqmi-proxy-1.32.0-1ubuntu0.22.04.1	amd64-deb	2/16/2024	
libsasl2-2-2.1.27+dfsg2-3ubuntu1.2	amd64-deb	2/16/2024	
libsasl2-modules-2.1.27+dfsg2-3ubuntu1.2	amd64-deb	2/16/2024	
libsasl2-modules-db-2.1.27+dfsg2-3ubuntu1.2	amd64-deb	2/16/2024	
libsasl2-modules-gssapi-mit-2.1.27+dfsg2-3ubuntu1.2	amd64-deb	Yesterday at 18:04	
libsgutils2-2-1.46-1ubuntu0.22.04.1	amd64-deb	2/16/2024	
libsmbclient-4.15.13+dfsg-0ubuntu1.6	amd64-deb	Yesterday at 18:04	
libssl3-3.0.2-0ubuntu1.16	amd64-deb	Yesterday at 16:30	
libsss-certmap0-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libsss-idmap0-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libsss-nss-idmap0-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
libsystemd0-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
libudev1-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
libunwind8-1.3.2-2build2.1	amd64-deb	2/16/2024	
libusb-1.0-0-1.0.25-1ubuntu2	amd64-deb	2/16/2024	
libwbclient0-4.15.13+dfsg-0ubuntu1.6	amd64-deb	Yesterday at 18:04	
libxtables12-1.8.7-1ubuntu5.2	amd64-deb	2/16/2024	
linux-firmware-20220329.git681281e4-0ubuntu3.31	all-deb	Yesterday at 16:26	
linux-generic-5.15.0.116.116-X	amd64-deb	Yesterday at 16:26	
linux-headers-5.15.0-116-5.15.0-116.126	all-deb	Yesterday at 16:26	
linux-headers-5.15.0-116-generic-5.15.0-116.126	amd64-deb	Yesterday at 16:26	
linux-headers-generic-5.15.0.116.116-X	amd64-deb	Yesterday at 16:26	
linux-image-5.15.0-116-generic-5.15.0-116.126	amd64-deb	Yesterday at 16:26	
linux-image-generic-5.15.0.116.116-X	amd64-deb	Yesterday at 16:26	
linux-modules-5.15.0-116-generic-5.15.0-116.126	amd64-deb	Yesterday at 16:26	
linux-modules-extra-5.15.0-116-generic-5.15.0-116.126	amd64-deb	Yesterday at 16:26	
mdadm-4.2-0ubuntu2	amd64-deb	2/16/2024	
modemmanager-1.20.0-1~ubuntu22.04.3	amd64-deb	2/16/2024	
motd-news-config-12ubuntu4.6-X	all-deb	2/16/2024	
multipath-tools-0.8.8-1ubuntu1.22.04.4	amd64-deb	2/16/2024	
nessusagent-10.5.1-X	amd64-deb	Yesterday at 17:12	
netplan.io-0.106.1-7ubuntu0.22.04.4	amd64-deb	Yesterday at 16:30	
nftables-1.0.2-1ubuntu3	amd64-deb	2/16/2024	
openssh-client-8.9p1-3ubuntu0.10	amd64-deb	Yesterday at 16:30	
openssh-server-8.9p1-3ubuntu0.10	amd64-deb	Yesterday at 16:30	
openssh-sftp-server-8.9p1-3ubuntu0.10	amd64-deb	Yesterday at 16:30	
openssl-3.0.2-0ubuntu1.16	amd64-deb	Yesterday at 16:30	
open-vm-tools-12.3.5-3~ubuntu0.22.04.1	amd64-deb	2/16/2024	
python3.10-3.10.12-1~22.04.4	amd64-deb	Yesterday at 16:30	
python3.10-minimal-3.10.12-1~22.04.4	amd64-deb	Yesterday at 16:30	
python3-apport-2.20.11-0ubuntu82.5	all-deb	2/16/2024	
python3-apt-2.4.0ubuntu3-X	amd64-deb	2/16/2024	
python3-debian-0.1.43ubuntu1.1-X	all-deb	2/16/2024	
python3-distro-info-1.1ubuntu0.2-X	all-deb	2/16/2024	
python3-distupgrade-22.04.19-X	all-deb	2/16/2024	
python3-gi-3.42.1-0ubuntu1	amd64-deb	2/16/2024	
python3-problem-report-2.20.11-0ubuntu82.5	all-deb	2/16/2024	
python3-software-properties-0.99.22.9-X	all-deb	2/16/2024	
python3-sss-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
python3-tz-2022.1-1ubuntu0.22.04.1	all-deb	2/16/2024	
python3-update-manager-22.04.18-X	all-deb	2/16/2024	
python-apt-common-2.4.0ubuntu3-X	all-deb	2/16/2024	
samba-libs-4.15.13+dfsg-0ubuntu1.6	amd64-deb	Yesterday at 18:04	
sg3-utils-1.46-1ubuntu0.22.04.1	amd64-deb	2/16/2024	
sg3-utils-udev-1.46-1ubuntu0.22.04.1	all-deb	2/16/2024	
software-properties-common-0.99.22.9-X	all-deb	2/16/2024	
sosreport-4.5.6-0ubuntu1~22.04.2	amd64-deb	2/16/2024	
sssd-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-ad-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-ad-common-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-common-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-dbus-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-ipa-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-krb5-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-krb5-common-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-ldap-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-proxy-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
sssd-tools-2.6.3-1ubuntu3.3	amd64-deb	Yesterday at 18:04	
systemd-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
systemd-hwe-hwdb-249.11.5-X	all-deb	2/16/2024	
systemd-sysv-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
systemd-timesyncd-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
tcpdump-4.99.1-3ubuntu0.1	amd64-deb	2/16/2024	
thermald-2.4.9-1ubuntu0.4	amd64-deb	Yesterday at 16:26	
ubuntu-advantage-tools-30~22.04-X	amd64-deb	2/16/2024	
ubuntu-drivers-common-0.9.6.2~0.22.04.6-X	amd64-deb	2/16/2024	
ubuntu-minimal-1.481.1-X	amd64-deb	2/16/2024	
ubuntu-pro-client-l10n-30~22.04-X	amd64-deb	2/16/2024	
ubuntu-release-upgrader-core-22.04.19-X	all-deb	2/16/2024	
ubuntu-server-1.481.1-X	amd64-deb	2/16/2024	
ubuntu-server-minimal-1.481.1-X	amd64-deb	2/16/2024	
ubuntu-standard-1.481.1-X	amd64-deb	2/16/2024	
udev-249.11-0ubuntu3.12	amd64-deb	2/16/2024	
ufw-0.36.1-4ubuntu0.1	all-deb	2/16/2024	
update-manager-core-22.04.18-X	all-deb	2/16/2024	
update-notifier-common-3.192.54.6-X	all-deb	2/16/2024	
wget-1.21.2-2ubuntu1.1	amd64-deb	Yesterday at 16:30	

```
i.e. the newest bind9 packages are no longer on the list (presumably because they aren't security updates in the security channels), all the packages which are newer versions than the packages in the base channels because they are bug-fix packages that aren't in the security updates channels.

For instance, the latest apport package in the security channel is apport-2.20.11-0ubuntu82.4.all-deb, older than the installed "Non-Compliant" apport-2.20.11-0ubuntu82.5 from the non-compliant list. The "Non-Compliant" sssd packages updates are not in the LCM security channel because they were released after the LCM cutoff date filter, so the installed "Non-Compliant" packages are newer than the stock 22.04 packages in the Main channel.

This behaviour does not match the one you claim we should see.

**Reply by rjmateus** [reply]
It looks like for the exact Name, Versio, Relaese, Epoc (if exist) of the packages that are installed and the packages that are available o the channels assigned in to the system.
In your last example, if the installed package is in a version higher then the one provided in the CLM channel, and the new version is not in the channel, it will show as non-compliant since the exact version is not available.

Regarding the package bind9 (for example `bind9-dnsutils`). From the vendor channel that contains all the packages, can you check all the versions that exists? For base channel and each child channels go to it's detail, click on packages, and then filter by `bind9-dnsutils`, then send back the list for each channel.

From information you have sent, looks like the versions are different:
- bind9-dnsutils-9.18.18-0ubuntu0.22.04.2
- bind9-dnsutils-9.18.24-0ubuntu0.22.04.1:1.amd64-deb

version 9.18.18 <> 9.18.24.

I did a quick search locally, and it looks like version 9.18.18 is only on channel `main-security`, and the version 9.18.24 is only on channel `main-updates`. When you assigned the non-CLM channels did you assign both main and security channels?
Another point is that some vendor only keep the latest version on the channel, and delete older version, this can lead packages to be non compliant if the exact version is not found.

**Reply by ppanon2022** [reply]
> It looks like for the exact Name, Versio, Release, Epoc (if exist) of the packages that are installed and the packages that are available o the channels assigned in to the system. In your last example, if the installed package is in a version higher then the one provided in the CLM channel, and the new version is not in the channel, it will show as non-compliant since the exact version is not available.
Correct. That's the previous behaviour and what I expected (although technically, if the package version is newer than what's available in the subscribed channels - because of an install-time general update that included more than security updates - it's arguable it should still be compliant). What I didn't expect is that packages in the Updates list would also show in the Non-Compliant list as well, which appears to be the case now and wasn't a few months ago.
> 
> Regarding the package bind9 (for example `bind9-dnsutils`). From the vendor channel that contains all the packages, can you check all the versions that exists? For base channel and each child channels go to its detail, click on packages, and then filter by `bind9-dnsutils`, then send back the list for each channel.
> 
> From information you have sent, looks like the versions are different:
> 
> * bind9-dnsutils-9.18.18-0ubuntu0.22.04.2
> * bind9-dnsutils-9.18.24-0ubuntu0.22.04.1:1.amd64-deb
> 
> version 9.18.18 <> 9.18.24.
> 

I think that's just because I pulled it from different pages and they display the packages differently but using the Software Package Search results show they're the same arch/naming scheme
Currently the LCM projects only include the security updates channel
bind9-dnsutils-9.18.18-0ubuntu0.22.04.2:1.amd64-deb | ubuntu22-prod-Ubuntu 22.04 LTS AMD64 Main Security for Uyuni
bind9-dnsutils-9.18.1-1ubuntu1:1.amd64-deb | ubuntu22-prod-Ubuntu 22.04 LTS AMD64 Main for Uyuni
bind9-dnsutils-9.18.18-0ubuntu0.22.04.2:1.amd64-deb | ubuntu22-it_baseline-Ubuntu 22.04 LTS AMD64 Main Security for Uyuni
bind9-dnsutils-9.18.1-1ubuntu1:1.amd64-deb | ubuntu22-it_baseline-Ubuntu 22.04 LTS AMD64 Main for Uyuni
bind9-dnsutils-9.18.18-0ubuntu0.22.04.2:1.amd64-deb | ubuntu22-devtest-Ubuntu 22.04 LTS AMD64 Main Security for Uyuni
bind9-dnsutils-9.18.1-1ubuntu1:1.amd64-deb | ubuntu22-devtest-Ubuntu 22.04 LTS AMD64 Main for Uyuni

But the actual sync channels include both the updates and security channels and you can see the naming scheme is the same
bind9-dnsutils-9.18.24-0ubuntu0.22.04.1:1.amd64-deb | Ubuntu 22.04 LTS AMD64 Main Updates for Uyuni
bind9-dnsutils-9.18.18-0ubuntu0.22.04.2:1.amd64-deb | Ubuntu 22.04 LTS AMD64 Main Security for Uyuni
bind9-dnsutils-9.18.1-1ubuntu1:1.amd64-deb | Ubuntu 22.04 LTS AMD64 Main for Uyuni

> I did a quick search locally, and it looks like version 9.18.18 is only on channel `main-security`, and the version 9.18.24 is only on channel `main-updates`. When you assigned the non-LCM channels did you assign both main and security channels? 
Yes, I assigned the updates channel (which implicitly includes all security updates in addition to bug fixes). The point is that update channel includes all the updates that are applied by the Ubuntu installer at update time, to avoid confusion in the Non-Compliant list from newer non-security updates applied at install-time.
>Another point is that some vendor only keep the latest version on the channel, and delete older version, this can lead packages to be non compliant if the exact version is not found.

Well, that might be a problem after you mainline the GSOC project to only download metadata and packages that are actually requested for installation. But historically I have a little trouble understanding how that would be an issue. Currently any packages in the metadata get downloaded/synced, and vendors who prune their package repo also prune the metadata. Uyuni also seems to keep packages depending on your sync settings, so even if the vendor purges a package from their repo, it doesn't mean Uyuni would (otherwise it shouldn't take three separate steps - deleting channels, packages and running spacewalk-data-fsck - to clean out an obsolete/EOL OS release). 

I suppose if you built a system, then waited a few days or weeks before joining it to Uyuni and in the meantime a new version was released and the older version was deleted, then you could conceivably have that happen. But you would have to have been unlucky enough to have multiple packages released between two daily sync jobs, and get one during installation that got deleted before the next daily sync. Even then I would think the package name/arch/packaging-format would still match even if that specific version wasn't in that repo. It would be a very unlikely use case that could be handled by matching on package name/arch/deb-vs.-rpm whereas that behaviour breaks much more common use cases.


The other thing we're seeing now that we weren't seeing before is that the update list now lists all versions of packages that are newer than the installed version. So, if 3 updates to a package have come out since the last time updates were applied, then all 3 updates now show up in the updates list. I guess the idea may have been so that if there was an intermediate version that included a security patch/CVE fix, you wouldn't miss seeing that implicitly included patch because only the latest version without that association was being displayed. The problem is that when you display all those versions, the admin can no longer easily choose to apply all those updates (and perhaps decide to hold back one or two in exceptional cases) because the client complains about duplicate packages. Instead, what should happen is that the latest package version should be associated (through a subquery) with all the CVE patches for versions that were released after the currently installed version. OK, that is somewhat more complicated since the selected systems may have a range of installed versions (you could still make it work by using the oldest version among the selected systems as your boundary condition), and you have to figure out some way to coalesce the associated patch results if there are multiple newer patches. More simply if there are multiple CVE patches from the different newer versions, you could just list the latest one. But it's still better than the current results where the patches will show up almost randomly associated with different versions for different packages when you have multiple versions of binary packages built from a single source package. The same patch can get listed in the update list with one version of a package and a different version of a fraternal package. It's a mess.

I suppose that the new scheduled auto-update feature is supposed alleviate the extra admin overhead and risk of human error in selecting packages, but that doesn't look like it would work on scenarios when you're using LCM projects in conjunction with more complicated maintenance schedules. The current update listing with multiple newer versions of a package are a step backwards as far as I'm concerned. Anything you select other than the latest version is just going to show as still needing to be updated after it's installed, so it doesn't make sense to show them as an update option to the administrator if they don't resolve the entry as a package needing an update.

**Reply by rjmateus** [reply]
We will look into this, since as far as I know the package upgrade list should show only the latest version available on the channels.

**Reply by ppanon2022** [reply]
Here's the list of held back packages (haven't done a zypper refresh in a while). Most of the held back packages appear to have newer versions in the OS repos than in the Uyuni repo so I figured it made more sense to stick with the Uyuni repo versions. This is still running 2024.05. We haven't upgraded to 2024.07 yet. 

```
ca07uyuni1:/srv/www/htdocs/pub/repositories/el/8/repodata # zypper update
Loading repository data...
Warning: Repository 'Update repository of openSUSE Backports' metadata expired since 2024-07-18 13:01:53 EDT.
Warning: Repository 'Main Update Repository' metadata expired since 2024-07-03 04:46:10 EDT.

    Warning: Repository metadata expired: Check if 'autorefresh' is turned on (zypper lr), otherwise
    manually refresh the repository (zypper ref). If this does not solve the issue, it could be that
    you are using a broken mirror or the server has actually discontinued to support the repository.

Reading installed packages...

The following 17 package updates will NOT be installed:
  apache-commons-csv apache-commons-ognl byte-buddy classmate cobbler concurrentlinkedhashmap-lru glassfish-fastinfoset glassfish-jaxb-runtime
  glassfish-jaxb-txw2 istack-commons-runtime jackson-module-jaxb-annotations jandex mybatis reprepro stax-ex xmlstreambuffer xsom
```

Hmm, I see that with 2024.07 repo-sync has changed to be strict sync and clear out old packages that are no longer in the Internet vendor masters. That is presumably why you wrote about packages disappearing when vendors only keep the latest version. Since we're still running 2024.05, that hasn't been a factor. That does bring up a question though, when the new repo-sync removes a package from the channel because it's been removed from the vendor's master repo, does the file get removed as if spacewalk-data-fsck had been run?

**Reply by ppanon2022** [reply]
Hmm. I thought that a 2024.08 release wasn't listed in the July community hours roadmap for releases and migration.

**Reply by ppanon2022** [reply]
Well, this whole discussion is now moot due to the switch to containerized Uyuni
